[AMD] Add cache modifier to BufferLoad and BufferStore

yiqian1 opened the pull request:
This patch adds cache modifiers to the buffer load and store operations.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/ConvertToBufferOps.cpp:
```diff
@@ -218,8 +215,8 @@ struct ConvertTritonStoreToBufferStore
 
-    if (op.getCache() != triton::CacheModifier::NONE)
-      return failure();
+    // if (op.getCache() != triton::CacheModifier::NONE)
```
Please delete instead of comment out.

antiagainst commented on third_party/amd/include/Dialect/TritonAMDGPU/IR/TritonAMDGPUOps.td:
```diff
@@ -109,13 +109,15 @@ def BufferLoadOp : TT_AMDGPU_Op<"buffer_load", [
       ins
       TT_Ptr:$ptr,
       I32Tensor:$offsets,
+      DefaultValuedAttr<TT_CacheModifierAttr, "::mlir::triton::CacheModifier::NONE">:$cache,
```
Do we have a mapping for all cache modifiers supported in the frontend? If not should we add verifier to only accept those we can handle?

giuseros replied:
From what I understand we can support all modifiers. Am I correct @yiqian1 ?

yiqian1 replied:
Yes, we can support all cache modifiers.

giuseros commented on third_party/amd/lib/TritonAMDGPUToLLVM/BufferOpsEmitter.cpp:
```diff
@@ -161,10 +164,53 @@ void BufferEmitter::fillCommonArgs(Type type, Value rsrcDesc,
-  // bits 1-2: SLC, DLC = 0 (similarly)
-  // bit 3: swizzled (0 for raw)
-  Value cacheModifiers = int_val(32, 0);
+  //            bit 0 = glc, bit 1 = slc, bit 2 = dlc (gfx10/gfx11),
```
I think this could be improved:
- This is the perfect place to explain what those names are (glc, slc, dlc, swc, scc, nt, sc1, sc0, scope). 
- gfx940 and MI300 I suppose are the same thing, right? 
- Why we have the table only for MI300? Does it mean that the `switch` only applies to MI300? Would this work for say MI200?

This is the way I would structure that code. 
We support mainly 4 architectures : gfx90a, gfx942, gfx11, gfx12 . 
I don't know how different they are, but in the worst case (where all those architectures are different) I would do something like:

```
Value getCacheModifierGfx94(triton::CacheModifier cm){
// bit 0 = sc0, bit 1 = nt, bit 3 = swz, bit 4 = sc1
// Glossary:
// nt means ....
// sc0 means ...
// swz means ....
// sc1 means ....
// Translation:
  // Op   | cm  | SC1 | SC0 | NT |
  // -----+-----+-----+-----+----+--
  // Load | .ca |  0  |  0  | 0  |
  //      | .cg |  0  |  x  | 1  |
  //      | .cs |  1  |  0  | x  |
  //      | .cv |  1  |  1  | x  |
  // -----+-----+-----+-----+----+--
  // Store| .wb |  1  |  0  | 0  |
  //      | .cg |  0  |  0  | 0  |
  //      | .cs |  1  |  0  | 1  |
  //      | .wt |  1  |  1  | x  |
  int32_t aux = 0;
  switch (cm) {
  case triton::CacheModifier::CA:
    aux = 0;
    break;
  case triton::CacheModifier::CG:
    if (isBufferLoad)
      aux |= 2;
    break;
  case triton::CacheModifier::CS:
    aux |= 16;
    if (!isBufferLoad)
      aux |= 2;
    break;
  case triton::CacheModifier::CV:
    aux |= 17;
    break;
  case triton::CacheModifier::WB:
    aux |= 16;
    break;
  case triton::CacheModifier::WT:
    aux |= 17;
    break;
  default:
    aux = 0;
  }
return i32_val(aux);
}
Value getCacheModifierWordForArch(triton::CacheModifier cm, std::string arch){
switch (arch){
case "gfx90xx":
return getCacheModifierGfx90(cm);
case "gfx94xx":
return getCacheModifierGfx94(cm);
case "gfx11xx":
return getCacheModifierGfx11(cm);
case "gfx12xx":
return getCacheModifierGfx12(cm);
default:
assert(); // or return i32_val(0)
}
}
Value cacheModifiers = getCacheModifierWordForArch(targetInfo.getArch()));
```

You don't have to do that at once for all the architectures. You can start with gfx940 (which is the most important for now) and then defer the rest (i.e., the rest will return 0). Also, if two archs share the same cache modifiers, you can have a single function (and accorporate the cases). 

My broader point is that we shouldn't leave this as black magic. We want to nail down what behaviour we enforce on AMD hardware with a given Triton cache modifier. It's hard to tell for normal loads, where we are forced to obey to the backend's rules, but here we can have very fine grained control on those modifiers.

guacamoleo replied:
@ramjana This is what Brian Sumner had concerns about in the "LLVM for Triton CodeGen" chat.

guacamoleo replied:
I'd like to partially redact this comment. I've been reading about the difference between these cache modifiers and memory consistency which are related but not the same. I think most of my concerns are addressed with this PR only addressing BufferLoads and not AtomicLoads. Therefore we are only concerned with matching performance on a best effort basis between the PTX spec and AMD hardware and not correctness. Sorry for the confusion.

So this looks fine for gfx942.

giuseros reviewed (commented):
It's looking good! I only would like that the translation `Triton Modifier -> AMD modifier` was documented in such a way that it's clear what behaviour we want to enforce on AMD with a given Triton cache modifier.

guacamoleo commented on third_party/amd/lib/TritonAMDGPUToLLVM/BufferOpsEmitter.h:
```diff
@@ -67,14 +67,16 @@ struct BufferEmitter {
   // Emit a predicated rocdl.raw.ptr.buffer.load
   Value emitLoad(Type type, Value rsrcDesc, Value offset, Value pred,
-                 Value falseVal);
+                 Value falseVal, CacheModifier cm);
```
Ah, I see here we're targeting rocdl.raw.ptr.buffer.load (https://mlir.llvm.org/docs/Dialects/ROCDLDialect/#rocdlrawptrbufferload-rocdlrawptrbufferloadop).

@ramjana You had created a table for comparing PTX to AMD cache modifiers. Do we have that for earlier architectures, and will we always be able to create a 1:1 mapping for future architectures?

yiqian1 replied:
I did refer to Raman's table while coding the cache modifiers. It'd be great to have a similar comparison for other architectures.

guacamoleo reviewed (changes requested):
Need confirmation that PR is only relating to cache performance and not correctness. I also agree that we need a cleaner interface to architecture specific mappings to keep future support clean.

yiqian1 commented:
Thank you, everyone, for your feedback and questions. I agree that we should use a placeholder for non-gfx942 targets for now and address the correct behaviors of cache modifiers later. I’ve spent quite some time trying to piece the information together, but I’m still uncertain about how cache modifiers are handled on non-gfx942 machines. It’s clear that different targets can use different cache modifiers, and the same cache modifier can correspond to different hardware behaviors. For now, I’ve decided to focus on gfx942 as a starting point. This should help us to dive deeper into understanding cache modifiers on all architectures.

giuseros commented on third_party/amd/lib/TritonAMDGPUToLLVM/BufferOpsEmitter.cpp:
```diff
@@ -33,6 +33,72 @@ bool isZero(Value v) {
   return false;
 }
+
+// Create the auxiliary/cachepolicy value of ROCDL::RawPtrBufferLoad/StoreOp
```
This is definitely looking nice! My feeling is that I would still like a bit more explanation:
- What does it mean a `{group,device,system,wave}` cache level? Are we enabling those systems? And why there are `x` in some cases?
- Maybe in `getCacheModifierForTarget` you can put a description of what `.ca`, `.cg`, `.cs`, etc... mean?

This is not a hard requirement for though, and I would also like other people to comment (cc @guacamoleo @sjw36 )

yiqian1 replied:
I'm not entirely sure how the cache levels (wave, group, device, system) factor into this, but the different combinations of the 3 bits seem to best align with the behavior of NVIDIA cache modifiers. Some bits are 'x' because their values do not affect the cache behavior.

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/BufferOpsEmitter.cpp:
```diff
@@ -33,6 +33,72 @@ bool isZero(Value v) {
+//      | .cs |  1  |  0  | 1  |
+//      | .wt |  1  |  1  | x  |
+int32_t getCacheModifierForCDNA3(triton::CacheModifier cm, bool isBufferLoad) {
+  int32_t aux = 0;
```
The magic numbers are hard to read. Can you do something like

```c++
const int sc0Bit = 0b1, ntBit = 0b10, sc1Bit = 0b1000;
...
case ...:
  aux |= ntBit | sc1Bit;
```

guacamoleo commented:
I met with Yi and we discuss a few things.
(1) We have similar cache modifier logic inside of https://github.com/triton-lang/triton/blob/main/third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp
We don't want to map modifiers (ca, cg...) to architecture bits (sc0, sc1, ...) in multiple locations. These both should be moved to a same unified location. @yiqian1 Can you work on combining these together?
(2) The in-code comment / documentation should include the recipe that we're using to convert ptx-centric modifiers to AMD architecture. MI300 ISA doc https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-mi300-cdna3-instruction-set-architecture.pdf Table 48. Load Controls maps scope of wave/group/device/system to sc/nt bits; we also need a recipe for converting ca/cg/cs/wt to wave/group/device/system so we can reliably reproduce this with every architecture (and understand it for this one). @yiqian1 and I will discuss this with Raman.

giuseros commented:
Hi @guacamoleo , @yiqian1 
I agree on basically everything you are discussing, thanks! This is really the direction I would like this to go!

guacamoleo commented:
@yiqian1 Let me know when you've written up that confluence page on cache modifiers.

yiqian1 commented:
I've updated the settings of control bits based on [this confluence page](https://confluence.amd.com/display/MLSE/Understand+cache+modifiers) and moved all cache-modifier related functions into the Utility file. Combining them into one function isn’t straightforward because they are in different dialects. Moving them into one file makes it easier for us to maintain them together.

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -356,6 +356,137 @@ void llStore(RewriterBase &rewriter, Location loc, Value ptr, Value val,
+int32_t
+getCtrlBitsForCacheModifierOnTarget(triton::CacheModifier cm, bool isBufferLoad,
+                                    mlir::triton::AMD::TargetInfo targetInfo) {
+  if (targetInfo.getGPUKind() == llvm::AMDGPU::GK_GFX942) // gfx942
```
We should check ISAFamily being CDNA3? With that we don't need the new `getGPUKind()`.

yiqian1 replied:
I'm not sure if all CDNA3 targets have the same control bit settings. From [this table](https://rocm.docs.amd.com/projects/llvm-project/en/latest/LLVM/llvm/html/AMDGPUUsage.html#amdgpu-amdhsa-memory-model-code-sequences-gfx940-gfx941-gfx942-table) Stores with cache modifiers may map to different control bits on gfx940/1 and gfx942. It needs further investigation. So this PR is just focusing on gfx942.

antiagainst replied:
wow. Okay that's fine then. Thanks for the explanation!

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.h:
```diff
@@ -48,6 +49,11 @@ void llStore(RewriterBase &rewriter, Location loc, Value ptr, Value val,
+// Get cache modifier information for creating load or store instruction
+std::pair<bool, bool> getCacheModifierFlagsForPredicatedCall(LLVM::CallOp);
+int32_t getCtrlBitsForCacheModifierOnTarget(triton::CacheModifier, bool,
+                                            mlir::triton::AMD::TargetInfo);
```
Can you use `const &` here?

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp:
```diff
@@ -103,10 +73,11 @@ class CallOpConversion : public OpRewritePattern<LLVM::CallOp> {
     //               | 1         | 0/1     | (wt) global store sc0 sc1
-    bool vialatileFlag = isPredicatedStoreWT(callOp);
-    bool nonTmpFlag = isPredicatedStoreCS(callOp);
+    bool volatileFlag, nonTmpFlag;
+    std::tie(volatileFlag, nonTmpFlag) =
```
`auto [volatileFlag, nonTmpFlag] =` would do?

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.h:
```diff
@@ -48,6 +49,11 @@ void llStore(RewriterBase &rewriter, Location loc, Value ptr, Value val,
              triton::CacheModifier cm = triton::CacheModifier::NONE);
 
+// Get cache modifier information for creating load or store instruction
+std::pair<bool, bool> getCacheModifierFlagsForPredicatedCall(LLVM::CallOp);
```
Could you also add comment to specify the two output `bool` values are?

guacamoleo commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -356,6 +356,137 @@ void llStore(RewriterBase &rewriter, Location loc, Value ptr, Value val,
+      mlir::LLVM::AMD::predicatedStoreWT);
+}
+
+// Utility function that returns flags <volatile, notemporal> for a predicated
```
spelling: nontemporal

guacamoleo commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -356,6 +356,137 @@ void llStore(RewriterBase &rewriter, Location loc, Value ptr, Value val,
+    aux |= sc1Bit;
+    break;
+  case triton::CacheModifier::WT:
+    aux |= 17;
```
Can you convert this magic number to have meaning like above.

guacamoleo commented on third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp:
```diff
@@ -356,6 +356,137 @@ void llStore(RewriterBase &rewriter, Location loc, Value ptr, Value val,
+  if (targetInfo.getGPUKind() == llvm::AMDGPU::GK_GFX942) // gfx942
+    return getCtrlBitsForCacheModifierOnGFX942(cm, isBufferLoad);
+  else
+    return getDefaultCtrlBitsForCacheModifier(cm);
```
Should we also emit a warning here that the user is requesting cache control which Triton isn't able to provide for the target architecture?
